src/transformers/models/mistral/modeling_mistral.py

```diff
@@ -1118,9 +1170,42 @@ def prepare_inputs_for_generation(
 
         # if `inputs_embeds` are passed, we only want to use them in the 1st generation step
         if inputs_embeds is not None and cache_position[0] == 0:
-            model_inputs = {"inputs_embeds": inputs_embeds}
+            model_inputs = {"inputs_embeds": inputs_embeds, "input_ids": None}
         else:
-            model_inputs = {"input_ids": input_ids.contiguous()}  # `contiguous()` needed for compilation use cases
+            # `contiguous()` needed for compilation use cases
+            model_inputs = {"input_ids": input_ids.contiguous(), "inputs_embeds": None}
+
+        if isinstance(past_key_values, StaticCache) and attention_mask.ndim == 2:
+            if model_inputs["inputs_embeds"] is not None:
+                batch_size, sequence_length, _ = model_inputs["inputs_embeds"].shape
+                device = model_inputs["inputs_embeds"].device
+            else:
+                batch_size, sequence_length = model_inputs["input_ids"].shape
+                device = model_inputs["input_ids"].device
+
+            dtype = self.lm_head.weight.dtype
+            min_dtype = torch.finfo(dtype).min
+
+            if isinstance(past_key_values, SlidingWindowCache):
+                sliding_window = (
+                    self.config.sliding_window if self.config.sliding_window is not None else sequence_length
+                )
+                target_length = max(sequence_length, sliding_window)
+            else:
+                target_length = past_key_values.get_max_length()
```

Comment:
I don't understand why we need to do this: if there is a sliding WindowCache it was init from the `config` and thus has a correct `sliding_window`. 

Then, `SlidingWindowCache. get_max_length` should take `sequence_length` as input to return the max and avoid having these checks here WDYT?

Reply:
Yes, it should but the PR for getting `max_length` on slidingWindow cache is on its way and not merged yet. 

So, to update you on our discussions with @gante which are currently in different PR comments: some cache classes now do not have a `max_length` (e.g. SlidingWindow). As commented in the code, sliding window technically has no max length and goes on a rolling basis. But in transformers what we want to check is the "maximum capacity of cache instance", independently of how cache handles new tokens going beyond that capacity.

So, in a different PR I changed naming to `get_max_cache_shape` which is more straightforward and added `get_max_cache_shape` for Sliding Window cache. We'll do a simple deprecation cycle, as we did for static cache's "max_batch_size". Until the linked PR is merged, I am copying this piece of code from `mistral` and using it in `phi3`. I have it noted and will handle it depending which one gets merged first :)

Reply:
alright sounds good. I just don't want us to add too much complexity to the code! 🤗